fish-rust/src/abbrs.rs

```diff
@@ -0,0 +1,479 @@
+#![allow(clippy::extra_unused_lifetimes, clippy::needless_lifetimes)]
+use std::{
+    collections::HashSet,
+    sync::{Arc, Mutex, MutexGuard},
+};
+
+use crate::{
+    wchar::L,
+    wchar_ffi::{WCharFromFFI, WCharToFFI},
+};
+use cxx::{CxxWString, UniquePtr};
+use once_cell::sync::Lazy;
+pub use widestring::{Utf32Str as wstr, Utf32String as WString};
+
+use crate::abbrs::abbr_ffi::abbrs_replacer_t;
+use crate::ffi::re::regex_t;
+use crate::parse_constants::SourceRange;
+
+use self::abbr_ffi::{abbreviation_t, abbrs_position_t, abbrs_replacement_t};
+
+#[cxx::bridge]
+mod abbr_ffi {
+    extern "C++" {
+        include!("re.h");
+        include!("parse_constants.h");
+
+        type SourceRange = crate::parse_constants::SourceRange;
+    }
+
+    enum abbrs_position_t {
+        command,
+        anywhere,
+    }
+
+    struct abbrs_replacer_t {
+        replacement: UniquePtr<CxxWString>,
+        is_function: bool,
+        set_cursor_marker: UniquePtr<CxxWString>,
+        has_cursor_marker: bool,
+    }
+
+    struct abbrs_replacement_t {
+        range: SourceRange,
+        text: UniquePtr<CxxWString>,
+        cursor: usize,
+        has_cursor: bool,
```

Comment:
can't put `usize` in UniquePtr as it is not supported, This can be removed once FFI code is removed.
Also address sanitizer was failing due to having null pointers, so I did not use them at other places also

Reply:
> can't put `usize` in UniquePtr

I think`SharedPtr<usize>` works. Your solution is fine too